### ISSUE TITLE
Proposal of simple cache system for files written to disk

### ISSUE THREAD

**Issue by Bentroen** [mcbeet/beet#150]
Hello!

While using `beet` in a project with 7k+ generated files, I've noticed that every time the project is built, all files are deleted and regenerated from scratch. While this is fine for small amounts of data, it can become very slow as the packs get filled with more and more content. For instance, a binary tree generated with `Context.generate.function_tree()` can easily consist of thousands of <1 KB files, which are slow to write to disk.

As most of the modifications done to a pack only involve a couple of files — especially when you're tweaking something very specific, such as a single position in a block model —, a caching system would avoid many unnecessary I/O operations, which could potentially make the build process much more efficient.

Below is a proposal for how the cache system could work:

<details>
<summary>Expand</summary>

# Idea

During each build, construct a `dict` where each key is a path to an output file relative to the build directory, and each corresponding value is the MD5 hash of that file¹.

This is how the process of writing to/retrieving from this cache would work:

- _[Build started]_
  - Read the cache into memory.
  - Create an empty `dict` to hold the new cache that will be saved at the end of this build.
  - For each file written to disk:
     - Look for the output path in the cache dict.
        - If it's **not** found, write the file.
        - If it **is** found and its hash matches the file being written, **don't write the file**. Remove this entry from the cache.
        - If it **is** found but the hash doesn't match the file being written, overwrite the file². Remove this entry from the cache.
     - Add the `path: hash` pair to the future cache.
  - If after writing all files there are still entries in the current cache, delete those files.
  - Write the "future cache" to disk.
- _[Build finished]_

The result is that, at the end of each build, we would have a  `dict` storing all the files that have been written to disk during this particular build, which the output of the next build could be checked against.

### Notes

¹ The hash could be taken differently according to the file type. For instance, PIL images could take the `Image.from_bytes()` method; JSON data can use string conversion (`json.dumps()`) and so on. As different file types will never have matching paths (since they have different extensions), this should be fine, as long as hashing methods are consistent _within each file type_.

² Deleting the file and writing it has the same practical effect as overwriting, and may be easier to achieve consistently, given that different file types are also written differently to disk.

## Considerations

- While this is not an ideal solution, as it doesn't avoid processing files that may end up never being written to disk, it's simple and less "intrusive" than other methods I could think of — such as inspecting plugins themselves to see if any code within it has changed —, so it looks to me like a good middle-ground solution.

- There's already a powerful caching system in place that could be taken advantage of. In the same fashion that the `.beet_cache` already contains `link` and `template` caches, a new cache type called `build`/`content` could be created for that purpose.

- It could be possible to disable the build cache with a special CLI command or setting in the config file, for small projects where the benefits of caching wouldn't outweigh the cost.

- The cache should ideally be cleared when linking to a Minecraft world with `beet link`, or otherwise changing the build directory, as it would require all files to be generated again.
    - Alternatively, the output path of the last build could be stored in a different cache entry, and checked at the start of each build; if it's different than the current path, then the cached data isn't used.
    - For the same reason, building with `zipped=True` would (probably) not benefit from the cache at all.

- Moving where a big bunch of files is located in the pack would still cause deletion/regeneration of identical files. As this is unlikely to happen very often, it's acceptable to have the files regenerate in this case, as detecting that those files have moved would be very advanced behavior, involving a lot more checks.

- An alternative approach would be to use the existing `beet watch` mechanic to copy only the files where modifications were detected. While this is _conceptually_ simpler, it seems much more difficult to implement, as altering a single source file may have a cascading effect over the output — tracing all changed files would be a daunting task. The advantage of implementing it at a lower level is that it's easier to detect duplicates, as at that point we already know _exactly_ what the file will look like.

### Benefits
- Gain of speed proportional to the number of files being written
- Simple and straight-forward; doesn't require "intrusive" techniques such as inspecting code
- Avoids many unnecessary I/O operations
- There's already a caching system in place that could be taken advantage of.

### Limitations
- Files that may end up never being written to disk will still be processed by `beet`
- Folders that were previously present but become empty (e.g. due to the output path of a given plugin being renamed) will remain in the build output, unless some check is put in place to delete them.
- The cache will become outdated if manual file modifications are done to the build folder, which it won't be aware of. This means the build process will no longer be _guaranteed_ to yield the "intended" output — the only way to achieve a 100% "clean" build would be to run `beet cache --clear` prior to building the project.

---
</details>

Now, I don't know if a feature like this is planned (or even if it's already in the works), nor if the way I described it would be suitable to implement. Though I'd love to hear some feedback on the idea, so tell me what you think =)

**Comment by vberlier** [comment]
First off, thanks a lot for the detailed proposal! I definitely think that this is something worth looking into. Beet currently allows plugins to implement their own caching, but yeah the assumption is that plugin authors would set it up themselves on a case-by-case basis to avoid doing work entirely. So right now caching is manual and usually done at a much earlier stage.

However, before diving into what could be done to optimize the output, I was curious about one thing. Beet is already kind of smart when it comes to filesystem operations. When your data pack contains files that are on the filesystem and that haven't been modified by the build process, beet will not read and parse the files from your source directory, just to write them as-is in the output directory. In this situation beet is smart enough to use `shutil.copyfile` instead which is supposed to be more efficient since it avoids loading the file into userspace buffers. The idea is that this pairs well with caching intermediate data packs. By saving data packs in the cache and then loading them into the beet context, all the cached files are layed out flat on the filesystem and therefore saving the pack only calls `shutil.copyfile`.

I decided to experiment a bit to see how much this actually saves. I added logs for showing the time spent in the build process (https://github.com/mcbeet/beet/releases/tag/v0.41.10). This plugin generates `37449` function files and caches the resulting data pack:

```python
# demo.py
def beet_default(ctx: Context):
    cache = ctx.cache[__name__]
    count = ctx.meta["count"]

    if cache.json.get("count") == count:
        ctx.data.load(cache.directory / "demo")
        return

    for node, function in ctx.generate.function_tree("demo:foo", range(count)):
        if node.partition(8):
            function.append(f"execute if score #index temp matches {node.range} run function {node.children}")
        else:
            function.append(f"execute if score #index temp matches {node.range} run say {node.value}")

    cache.json["count"] = count
    ctx.data.save(path=cache.directory / "demo", overwrite=True)
```

```yml
# beet.yml
output: "out"
pipeline:
  - demo
meta:
  count: 100000
```

On the first run I get this:

```bash
$ beet --log debug
Building project...

DEBUG  | time  Pipeline took 3.21s.
DEBUG  | time  Output took 2.11s.
DEBUG  | time  Linking took 0.00s.

Done!
```

In the first run, the plugin needs to save the data pack twice, once during the pipeline to put it into the cache, and once after the build to save it in the output directory. This means that we can estimate that generating the tree takes roughly 1 second and saving the data pack normally takes 2 seconds.

And this is what the timings look like when it uses the cache:

```bash
$ beet --log debug
Building project...

DEBUG  | time  Pipeline took 1.98s.
DEBUG  | time  Output took 3.00s.
DEBUG  | time  Linking took 0.00s.

Done!
```

I was kind of surprised by the results. On the second run the plugin loads the pack from the cache during the pipeline, and then after the build it saves the pack using `shutil.copyfile`. It's interesting to note that in this particular example the function tree generator is quite fast. Loading the cached data pack takes about 2 seconds whereas generating it only took 1 second in the first run. This is most likely due to the fact that loading the pack ultimately requires a bunch of system calls to walk the directory. But because the plugin didn't need to save the intermediate pack into the cache the pipeline still runs 1 second faster.

For the output there isn't anything special going on, the second run saves the pack with the supposedly faster method but takes 3 seconds instead of 2 when the data pack is saved by simply writing each file. Looks like copying a lot of small files is more expensive than just plain writing. I think `shutil.copyfile` really shines when you have large files, and in this case since the files are small the writes are probably really cheap. So overall it looks like caching large data packs doesn't make that much of a difference between the first and the subsequent runs when you're not in a cpu-bound scenario. Also, this confirms that with large enough data packs caching has non-negligible overhead, so if the performance problem comes from IO, then the cache just adds even more IO, and in the example it increases the total build time to 5 seconds compared to the original 3 seconds.

TL;DR: If you're just generating a ton of files really quickly the current cache won't help you improve the output performance and can even make it worse by adding more IO.

Anyway, this wasn't directly about your proposal but it gave me a better understanding of the current limitations of the cache. So with all that in mind, investigating some sort of output cache makes a lot of sense. I think hashing everything can come with its own problems, and as you said it would make it tricky to remove old files and directories from the output. But it's definitely worth a shot. I'll try to experiment a bit with the idea. Also maybe you could try running your project with the latest release to confirm that it's the output phase that's taking the most time.

**Comment by Bentroen** [comment]
Thank you for the detailed write-up! It's really interesting to see how everything we discussed relates to my project. You can see I still have a lot to learn about how file handling works, haha!

As you pointed out, my initial impression that most of the time came from writing files repeatedly was mostly wrong. After having the output take so long, my mind immediately went to I/O. However, I completely flew past the things that _are_ already there, such as storing the files as unloaded, deserialized, or serialized. A small amount of knowledge about it would show there was a lot of potential for optimization there (as shown by my tests).

As you said, profiling could have ruled out that possibility right from the start, which would disprove my assumption that I/O operations were taking the longest. It wasn't a coincidence that the cache saved so much time; it was conceived exactly for this purpose and I just had not been using it properly. (By the way, I'm truly sorry if at any moment I made it sound like that the current cache system is not good; when used properly it can be insanely powerful!)

Now, this whole thing seems like a solution to a problem that isn't there hahaha. A hashing approach would make file writing clunkier to make up for the lack of optimization that _should_ have been made at a much higher level — like fitting a square peg in a round hole.

While my tests show that the potential gain of speed would be minimal, I'd still like to experiment with the idea, just to see what the results are. As you mentioned, hashing isn't a cheap operation either; doing it for thousands of files, even though most are very small, could turn out to be much slower. Also, it would require unloaded files to be converted to their serialized state, which is by no means good. Maybe in the end it would take much longer than the current approach? (perhaps I'll try that in my own free time :) )

Even if in the end it magically cuts off a few seconds from the writing time, I'm no longer sure it would be a welcome addition to `beet` — as you said, it has its own set of problems, and it's not the bottleneck anyway. A better idea would be to tackle down why the serialization is taking 62 seconds — ~I'll make sure to share my results on Discord so we can figure that out.~ Hold on a sec...

> Also if the caching becomes the bottleneck, it would be worth trying to save the intermediate data pack as a zip file, to avoid having to go through the kernel to walk over the directories when loading it from the cache, or saving the final data pack as a zip file, or both.

This is really interesting. Once again, my initial thought was that zip files would always take longer to save than disk files, since zip compression is not a cheap operation either. But after everything we discussed, it makes sense that it'd take less long, as it'd avoid many of the individual file copy operations.

So, I decided to test this. Except this time I removed a component that wasn't being cached on the first tests, as to make the results more consistent.

![image](https://user-images.githubusercontent.com/29354120/138382351-8ab157bd-b24f-42d6-8056-000343848f35.png)

(I repeated each test multiple times to ensure the times were accurate).

Immediately, the results are pretty interesting. The two scenarios where the intermediate pack was NOT zipped were the slowest, and, for some reason, in both of them the second run seems to take _longer_ than the first, until they "stabilize" the third time. Again, my starting assumption was wrong — zipping both the intermediate and the output pack yields the fastest results, building it consistently in about two seconds! I couldn't believe my eyes, as much that I had to go check if the files were really in the pack.

More importantly, this shows, once and for all, that writing is indeed not the problem! Speeding up the build process was a matter of using the right settings combined with an adequate setup of the cache, and I really doubt it could be sped up more than this, except maybe for the cases where the output isn't zipped.

However, this still fails to explain why was I getting linking times of 70 seconds before, even though I was using the slowest option (neither saved as zip).

To see how `beet` would perform mixing and matching files from different sources, this time I deliberately left a single file to _always_ be generated instead of being pulled from the cache. (Remember that component I had removed before?)

```shell
DEBUG  | time  Pipeline took 78.22s.
DEBUG  | time  Output took 0.00s.
DEBUG  | time  Linking took 7.07s.
```

```shell
DEBUG  | time  Pipeline took 5.16s.
DEBUG  | time  Output took 0.00s.
DEBUG  | time  Linking took 72.43s.
```

Whoa! This single file added 60 seconds to the build time?!

Well, this particular file is a `Structure` which happens to put almost a million `dict`s into memory. While I knew from the start that this wasn't very memory-efficient, it didn't seem too bad in terms of processing time, since the "Collision generated" was reported to be very fast.

But with your explanation about serialized and deserialized files, it all starts to make sense. On my previous tests, at the point the functions from the function tree were cached, the structure data was already in the pack, but _deserialized_! So once the functions were saved to the cache, `beet` was forced to convert it to the serialized state, which turns out was the one operation responsible for the bottleneck.

Problem solved! Ahem, excuse me, I'll go bury my head in the ground 😅

But seriously, thank you so much for your time! I learned a lot from reading your explanation, and I'll make sure to experiment more with the solutions you have suggested. Feel free to close this issue whenever you want :)

**Comment by vberlier** [comment]
Oh wow! What an absolutely fantastic write-up!

This chart is really insightful. With this whole conversation I think I can understand each timing now, thanks a lot for putting in the work!

I think the time increase on the second run when the intermediate data pack is not zipped probably has to do with the fact that the files we're reading were just being written to. I think operating systems can optimize "hot" files for either writing or reading, and in the second run it would probably need to undo these optimizations for each file to make them "hot" for reading instead. Afterwards since the files are "hot" for reading the build time stabilizes.

Also I'm not surprised that the main culprit was a structure file. `nbtlib` isn't too happy when parsing or serializing really large files. Luckily this is something that you can easily cache on its own.

I agree that after all this the output cache doesn't seem like it would bring sufficient benefits to be worth it. It's good that this led us to investigate though. I'll close the issue for now.

Once again thanks for really digging deep into this!
